CodeCoverage: HTML template facelift

milo opened the pull request:
Reworking #237, colors by GitHub.

Colors are not 1:1 with GitHub, it would need better granularity than `highlight_string()`, tokenize. Anyway, I'm not sure we want the look same as GitHub. When I had opened more windows, I'm loosing focus what is a code coverage and what is the GitHub.

But it is not about colors only. After this refactoring, output HTML is 752kB instead of 1.23MB, so maybe worth it.

Before some color fight, post some CSS propose please, e.g. on fiddle, if possible.

Example is here: https://hoola.cz/files/coverage.288-1.html

Mikulas commented:
Definitely an improvement over the current style, so for that +1.

Are overflows intentionally overflowing or is it an oversight? It's probably better to show the code rather than cropping it to the box, but it does not look that great. Forced scrolling might be a viable option (which is what github [does](https://github.com/nette/tester/blob/master/src/CodeCoverage/Generators/template.phtml)).

<img width="1440" alt="screenshot 2016-02-12 23 18 55" src="https://cloud.githubusercontent.com/assets/192200/13022061/2ab67606-d1df-11e5-9484-7a53fb13dd9b.png">
(example of `CodeCoverage\Generators\template.phtml`)

For the record, I don't really care about output size, not of the built html anyway.

Proposal: clicking a file should actually change fragment in url to simplify sharing between coworkers. Similarly so should line numbers. This could be sent in a future PR though.

dg commented:
Isn't it better to use color scheme similar to Tester API doc? https://api.nette.org/tester/1.7.0/source-Framework.Assert.php.html

milo commented:
@Mikulas Overflow was intentional.

@dg Maybe it would look better, but when trying it, there is almost everything `default` or `keyword` by `highlight_string()`. And it looses clarity.

milo commented:
I'm thinking about tokenizing, but have to measure how slowet it would be.

milo commented:
It tried own highlighter. It is about 31% slower, mainly due to tokenize.
- with [GitHub theme](https://hoola.cz/files/coverage.288-2.html)
- with [Nette ApiGen colors](https://hoola.cz/files/coverage.288-3.html)

dg commented:
Wow, own highlighter!

dg commented:
I am perfecty fine with GitHub theme and with Nette theme too. For consistency and in order to not look like GitHub rip, I am more for Nette colors. And links (filenames) should match palette too.

I see major importance of highlighter in differentiate between tested lines / untested lines / class & method declaration / comments. Everything other like highlight of variables etc is actually rather disturbing. So default `highlight_string()` with only 'default' and 'keyword' is probably for this case adequate.

milo commented:
> Wow, own highlighter!

But still, it needs more granularity! ;) It is just test digged out from one of my old lib. I'm planning to revert it.

> Everything other like highlight of variables etc is actually rather disturbing

Have to admit. I'm using it whole the day and it is really distracting.

milo commented:
What about [this approach](https://hoola.cz/files/coverage.288-4.html)?

dg commented:
Nice!

dg commented:
Overflow seems like a bug, especially now when code has bigger font, why is it intentional?

milo commented:
Well, I dont like vertical scrolling to much. But now it looks better with scroll bars 
- https://hoola.cz/files/coverage.288-5.html

milo commented:
Rebased and squashed. Imho ready to merge.
- [before](https://hoola.cz/files/coverage.html) 1243kB
- [now](https://hoola.cz/files/coverage.288-6.html) 891kB

dg commented on src/CodeCoverage/Generators/template.phtml:
```diff
@@ -98,49 +134,39 @@
 			<td class="number"><small><?= $info->coverage ?> %</small></td>
 			<td><div class="bar"><div style="width: <?= $info->coverage ?>%"></div></div></td>
-			<td><a href="#fragment<?= $id ?>" onclick="toggle(this); return false"><span><?= $info->name ?></span></a></td>
+			<td><a href="#F<?= $id ?>" onclick="toggle(this.href); return false"><span><?= $info->name ?></span></a></td>
```
`<span>` around `$info->name` seems useless

dg commented on src/CodeCoverage/Generators/template.phtml:
```diff
@@ -9,45 +9,72 @@
+
+	code,
+	div.code {
+		font-family: monospace;
```
IMHO something like `font: 13px/1.3 monospace;` should make it better.

dg commented:
There is problem with short hightlighted lines, when you scroll code right. The solution is:

```html
<div class="code">  (remove here position:relative, it will become useles)
NEW   <div class="content"> // float: left; position: relative
NEW      <div class="lines">  // min-width: 100%; position: absolute;
                    <div><a href='#F1L1' id='F1L1'> 1</a>…  //original divs with lines
              </div>
              <code> ..... </code>   // position: relative (instead of absolute), left padding 3em (instead of left: 3em), maybe display: block
          </div>
</div>
```

milo commented:
@dg First things fixed. But I'am not able to handle the short lines issue. There is always some ugly margin or white-space issue. Frontend help is very appreciated :) Or I'll elaborate later.

dg commented on src/CodeCoverage/Generators/template.phtml:
```diff
@@ -176,18 +176,21 @@
+				el.style.display = 'block';
+			}
+			e.preventDefault();
 			return false;
```
return false can be removed

dg commented on src/CodeCoverage/Generators/template.phtml:
```diff
@@ -176,18 +176,21 @@
 	if (window.location.hash.length) {
-		toggle(window.location.hash.split('L')[0]);
+		document.getElementById(window.location.hash.replace(/^#|L\d+$/g, '')).style.display = 'block';
 	}
```
more error prone `if (el = document.getElementById(window.location.hash.replace(/^#|L\d+$/g, ''))) { el.style.display = 'block'; }`

dg commented:
Ok, merge it and I'll try to solve line issue :-)

milo commented:
> I'll try to solve line issue

Thank you! Next beer is on me :)

dg commented:
:beers:
